### ISSUE TITLE
to_dict() infinite recursion

### ISSUE THREAD

**Issue by adm271828** [dgilland/sqlservice#27]
Given the following code:

```python
from sqlalchemy import MetaData
from sqlalchemy import Column, ForeignKey, orm
from sqlalchemy import Integer

from sqlservice import SQLClient
from sqlservice import ModelBase, as_declarative
  
metadata = MetaData()

@as_declarative(metadata = metadata)
class Model(ModelBase):
  pass

config = {
  'SQL_DATABASE_URI':      'sqlite:///:memory:',
}

db = SQLClient(config, model_class = Model)

class A(Model):
  __tablename__ = 'a'

  id = Column(Integer, primary_key=True)
  r0 = orm.relationship('B', back_populates = 'r1', lazy = 'joined')
  
  def foo(self):
    self.r0.append(B(i1 = self.id))
    
class B(Model):
  __tablename__ = 'b'

  id  = Column(Integer, primary_key = True)
  i1 = Column(Integer, ForeignKey('a.id'))
  r1 = orm.relationship('A', back_populates = 'r0', uselist = False)
```

Create an instance of `A` and append some `B`.

```python
In [2]: db.create_all()

In [3]: a0=A()

In [4]: a0
Out[4]: <A(id=None)>

db.save(a0)
Out[5]: <A(id=1)>

In [6]: a0
Out[6]: <A(id=1)>

In [7]: a0.foo()

In [8]: db.session.commit()

In [9]: a0
Out[9]: <A(id=1)>

In [10]: a0.r0
Out[10]: [<B(id=1, i1=1)>]

In [11]: a0.to_dict()
Out[11]: {'id': 1, 'r0': [{'i1': 1, 'id': 1}]}
```

Try again, and call `to_dict()` before commiting.

```python
In [12]: a0.foo()

In [13]: a0.r0
Out[13]: [<B(id=1, i1=1)>, <B(id=None, i1=1)>]

In [14]: a0.to_dict()

  ....
  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 216, in to_dict
    data[key] = adapter(value, key, self)

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 324, in default_dict_adapter
    for val in value]

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 324, in <listcomp>
    for val in value]

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 216, in to_dict
    data[key] = adapter(value, key, self)

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 320, in default_dict_adapter
    value = value.to_dict()

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 216, in to_dict
    data[key] = adapter(value, key, self)

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 321, in default_dict_adapter
    elif is_sequence(value):

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\utils.py", line 71, in is_sequence
    (isinstance(value, Sequence) and not isinstance(value, str)))

  File "C:\Tools\Continuum\Anaconda3\lib\abc.py", line 190, in __instancecheck__
    subclass in cls._abc_negative_cache):

  File "C:\Tools\Continuum\Anaconda3\lib\_weakrefset.py", line 75, in __contains__
    return wr in self.data

RecursionError: maximum recursion depth exceeded in comparison
```
But if we commit before, this is OK:
```python

In [15]: a0.r0
Out[15]: [<B(id=1, i1=1)>, <B(id=None, i1=1)>]

In [16]: db.session.commit()

In [17]: a0.to_dict()
Out[17]: {'id': 1, 'r0': [{'i1': 1, 'id': 1}, {'i1': 1, 'id': 2}]}
```

Maybe there is something wrong in my model definition. But where? Or maybe something wrong with `to_dict()`?

Best regards,

Antoine

**Comment by adm271828** [comment]
Even after commit. Seems to be related to the bidirectional relationship.

```python
In [2]: db.create_all()

In [3]: a0=A()

In [4]: db.save(a0)
Out[4]: <A(id=1)>

In [5]: a0.foo()

In [6]: a0.r0
Out[6]: [<B(id=None, i1=1)>]

In [7]: db.commit()

In [8]: a0.r0
Out[8]: [<B(id=1, i1=1)>]

In [9]: a0.r0[0].r1
Out[9]: <A(id=1)>

In [10]: a0.r0[0].r1.r0
Out[10]: [<B(id=1, i1=1)>]

In [11]: a0.to_dict()

  ....
  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 216, in to_dict
    data[key] = adapter(value, key, self)

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 324, in default_dict_adapter
    for val in value]

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 324, in <listcomp>
    for val in value]

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 216, in to_dict
    data[key] = adapter(value, key, self)

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 320, in default_dict_adapter
    value = value.to_dict()

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 216, in to_dict
    data[key] = adapter(value, key, self)

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\model.py", line 321, in default_dict_adapter
    elif is_sequence(value):

  File "C:\Tools\Continuum\Anaconda3\lib\site-packages\sqlservice\utils.py", line 71, in is_sequence
    (isinstance(value, Sequence) and not isinstance(value, str)))

  File "C:\Tools\Continuum\Anaconda3\lib\abc.py", line 190, in __instancecheck__
    subclass in cls._abc_negative_cache):

  File "C:\Tools\Continuum\Anaconda3\lib\_weakrefset.py", line 75, in __contains__
    return wr in self.data

RecursionError: maximum recursion depth exceeded in comparison
```

**Comment by dgilland** [comment]
This is a known issue where fully loaded joined models with references to each other result in each model serializing the other which contains a reference back to itself. The reason it sometimes doesn't result in the maximum recursion error has to do with whether the joined model is loaded or not (e.g. after a commit, SQLAlchemy marks data as "stale" so that the next access to that data will result in a database lookup). The way that `model.to_dict()` works is that it will only load data that doesn't result in a database lookup (so as to avoid N+1 query issues) which basically means only looking at the data in `model.__dict__`.

The general recommendation is to override `to_dict()` or set dict adapters that prevent one side from loading the relationship data or use a custom serializer (outside of the `Model.to_dict()` scope) to avoid the recursion.

**Comment by adm271828** [comment]
Thanks for the explanation. The ability to have a `null` adapter would help in this case.

Best regards.

**Comment by dgilland** [comment]
Additional workaround now available in [v1.1.3](https://pypi.org/project/sqlservice/1.1.3/) as described in #26.

**Comment by Zaffer** [comment]
could one use sqlalchemy_serializer's SerializerMixin to help with this?
